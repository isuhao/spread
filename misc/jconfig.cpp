#include "jconfig.hpp"
#include "readjson.hpp"
#include "comp85.hpp"
#include <stdexcept>
#include <boost/thread/recursive_mutex.hpp>
#include <boost/filesystem.hpp>

using namespace Misc;

//#define PRINT_DEBUG
#ifdef PRINT_DEBUG
#include <iostream>
#define PRINT(a) std::cout << a << "\n"
#else
#define PRINT(a)
#endif

namespace bs = boost::filesystem;

static void parent(const bs::path &file)
{
  bs::create_directories(file.parent_path());
}

struct JConfig::_JConfig_Hidden
{
  Json::Value val; 
  boost::recursive_mutex mutex;

  Json::Value LG(const std::string &name, const Json::Value &v)
  {
    mutex.lock();
    Json::Value res = val.get(name, v);
    mutex.unlock();
    return res;
  }
};

#define LOCK boost::lock_guard<boost::recursive_mutex> lock(p->mutex)
#define WLOCK assert(!readOnly);LOCK

JConfig::JConfig(const std::string &_file, bool _readOnly)
  : readOnly(_readOnly), file(_file)
{
  p = new _JConfig_Hidden;
  load();
}
JConfig::~JConfig() { delete p; }

/* Mutex notes:

   The mutex is intended primarily to protect the setter and getter
   functions to avoid multiple simultaneous access to the Json::Value
   or the config file.
 */

void JConfig::load()
{
  // We never fail. A missing or invalid file is OK.
  if(file == "") return;

  LOCK;
  try
    {
      std::string toLoad = file;

      // If the file does not exist, but .old does, revert to the
      // transactional backup.
      if(!bs::exists(toLoad))
        {
          toLoad += ".old";
          if(!bs::exists(toLoad))
            return;
        }
      p->val = ReadJson::readJson(toLoad);
    }
  catch(...) {}
}

void JConfig::save()
{
  if(file == "")
    return;

  std::string newFile = file + ".new";
  std::string oldFile = file + ".old";

  WLOCK;

  // First, write to .new
  parent(newFile);
  ReadJson::writeJson(newFile, p->val);

  // Then rename existing file (if any) to .old
  if(bs::exists(file))
    {
      if(bs::exists(oldFile))
        bs::remove(oldFile);
      bs::rename(file,oldFile);
    }

  // Finally, move .new into place
  bs::rename(newFile, file);
}

void JConfig::setBool(const std::string &name, bool b)
{ WLOCK; p->val[name] = b; save(); }

void JConfig::setInt(const std::string &name, int i)
{ WLOCK; p->val[name] = i; save(); }

void JConfig::set(const std::string &name, const std::string &value)
{
  PRINT("Conf: Setting " << name << " => " << value);
  WLOCK;
  p->val[name] = value;
  save();
}

bool JConfig::getBool(const std::string &name, bool def)
{ return p->LG(name,def).asBool(); }

int JConfig::getInt(const std::string &name, int def)
{ return p->LG(name,def).asInt(); }

void JConfig::setInt64(const std::string &name, int64_t i)
{ setData(name, &i, 8); }

int64_t JConfig::getInt64(const std::string &name, int64_t def)
{
  int64_t res;
  try { getData(name, &res, 8); }
  catch(...) { res=def; }
  return res;
}

bool JConfig::remove(const std::string &name)
{
  WLOCK;
  // removeMember() returns a Null value if nothing was removed.
  bool res = !p->val.removeMember(name).isNull();
  save();
}

void JConfig::setMany(const std::map<std::string,std::string> &entries)
{
  if(entries.size() == 0) return;

  WLOCK;
  std::map<std::string,std::string>::const_iterator it;
  for(it = entries.begin(); it != entries.end(); it++)
    p->val[it->first] = it->second;
  save();
}

std::string JConfig::get(const std::string &name, const std::string &def)
{ return p->LG(name,def).asString(); }

bool JConfig::has(const std::string &name)
{ return p->val.isMember(name); }

std::vector<std::string> JConfig::getNames()
{ return p->val.getMemberNames(); }

void JConfig::setData(const std::string &name, const void *p, size_t num)
{
  set(name, Comp85::encode(p,num));
}

void JConfig::getData(const std::string &name, void *p, size_t num)
{
  std::string val = get(name);
  if(Comp85::calc_binary(val.size()) != num)
    throw std::runtime_error("Space mismatch in field '" + name + "'");
  Comp85::decode(get(name), p);
}

