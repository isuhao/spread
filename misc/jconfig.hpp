#ifndef __JCONFIG_HPP_
#define __JCONFIG_HPP_

#include <string>
#include <vector>
#include <map>
#include <stdint.h>

namespace Misc
{
  /* This struct reads and writes config to a json file.
   */
  struct JConfig
  {
    bool readOnly;

    JConfig(const std::string &_file = "", bool _readOnly=false);
    ~JConfig();

    void load();
    void save();

    void load(const std::string &_file)
    { file = _file; load(); }
    void save(const std::string &_file)
    { file = _file; save(); }

    void setBool(const std::string &name, bool b);
    bool getBool(const std::string &name, bool def=false);

    void setInt(const std::string &name, int b);
    int getInt(const std::string &name, int def=0);

    // Sets 64 bit data. Uses binary encoding, since jsoncpp doesn't
    // reliably support 64 bit sizes.
    void setInt64(const std::string &name, int64_t i);
    int64_t getInt64(const std::string &name, int64_t def=0);

    // Store binary data encoded in a string
    void setData(const std::string &name, const void *p, size_t num);
    void getData(const std::string &name, void *p, size_t num);

    // Remove an entry. Non-existing entries are allowed. Returns true
    // if the entry existed before removal.
    bool remove(const std::string &name);

    // Set or get string entries
    void set(const std::string &name, const std::string &value);
    std::string get(const std::string &name, const std::string &def="");

    /* Mass set string entries. Use this if you are writing large
       amounts of data.

       JConfig rewrites the file after every single set* operation.
       This is usually OK for low-traffic config files. For some uses
       however, a file may have many thousands of entries. Rewriting a
       config file of several megabytes thousands of times is highly
       impractical.

       This function lets you set any number of entries (currently as
       strings only) in one operation, and thus only perform one
       single file write.
     */
    void setMany(const std::map<std::string,std::string> &entries);

    bool has(const std::string &name);

    std::vector<std::string> getNames();

  private:
    std::string file;
    struct _JConfig_Hidden;
    _JConfig_Hidden *p;
  };
}

#endif
