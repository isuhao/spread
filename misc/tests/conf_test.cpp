#include "jconfig.hpp"

#include <iostream>
#include <assert.h>

using namespace std;
using namespace Misc;

void test1()
{
  cout << "Writing to fresh object:\n";
  JConfig jc;
  jc.setBool("am_i_beautiful", true);
  jc.setBool("am_i_an_umbrella", false);

  cout << "  Beautiful: " << jc.getBool("am_i_beautiful") << endl;
  cout << "  Umbrella: " << jc.getBool("am_i_an_umbrella") << endl;
  cout << "  Spork: " << jc.getBool("spork") << endl;
  cout << "  A cloud: " << jc.getBool("cloud", true) << endl;

  jc.save("_test1.conf");

  JConfig jc2("_test1.conf");
  assert(jc2.getBool("am_i_beautiful"));
}

void test2()
{
  cout << "Writing to pre-loaded object:\n";
  JConfig jc("_test2.conf");

  jc.setInt("morning flower", 33);

  cout << "  Morning: " << jc.getInt("morning flower", 19) << endl;
  cout << "  Evening: " << jc.getInt("evening evil", 19) << endl;

  JConfig jc2("_test2.conf");
  assert(jc2.getInt("morning flower") == 33);
}

struct Tmp
{
  float f;
  int i;

  Tmp() : f(0), i(0) {}
};

void test3()
{
  cout << "Testing binary data:\n";
  JConfig jc("_test3.conf");

  Tmp t1, t2;
  t1.f = 3.14;
  t1.i = 103;
  jc.setData("fval", &t1, sizeof(t1));
  jc.getData("fval", &t2, sizeof(t2));

  cout << "  f: " << t2.f << endl;
  cout << "  i: " << t2.i << endl;
}

int main()
{
  test1();
  test2();
  test3();

  return 0;
}
