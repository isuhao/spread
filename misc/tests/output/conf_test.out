Writing to fresh object:
  Beautiful: 1
  Umbrella: 0
  Spork: 0
  A cloud: 1
Writing to pre-loaded object:
  Morning: 33
  Evening: 19
Testing binary data:
  f: 3.14
  i: 103
